### ISSUE TITLE
List of topics to cover from discussion so far.

### ISSUE THREAD

**Issue by mhdawson** [nodeshift/web-application-reference#1]
Frameworks
Evaluating frameworks and tools - **ACTION: @bartpotts**
Level of Abstraction: more abstract frameworks can cover over cumbersome platform APIs. Future proofing - avoiding “The Big Rewrite”. Tradeoffs: highly abstract frameworks can capture platform
Optimize for the right things: pick frameworks which emphasize end user outcomes and maintainability
Performance: benchmarks are fun to pore over, but seek out case studies of major performance struggles and how they were overcome

- Web Standards
  - Can I use ?
  - Transpilers (support for stage 3 stuff)
  - Web Components
  - Level of impact is affected by internal/external site visibility
- How you compile create the front end JavaScript
- Deployment - **ACTION: @lholmquist** 
	- Nginx to deploy
	- Edge Deployment
- Server Components/Streaming HTML
- Other rendering methods **ACTION: @tomev10**
  - The methods
    - Static Site Generated (SSG)
    - Server Side Rendering (SSR)
    - Client Side Rendering (CSR)
  - The pros and cons of each on
    - SEO
    - Hydration
    - Requirements on JS
    - Server Compute
    - Dynamic content
- Re-use
  - Micro front ends
  - Web components
  - DRY
- Build tools
  - Bundling
  - Minifying
  - Obfuscation
  - Transpilation
  	- Babel
  - Typescript
  - Supporting multiple browsers
  - Polyfills
  - Chunking strategies
  - Module patterns (ESM, common JS, RequireJS, AMD, UMD)
- Styles
  - CSS
  - Style-Components
  - StyleSheets
  - Inline Styles
  - CSS in JS
  - SASS/SCSS/LESS
  - Templates?
- Static assets
- Authentication
- Security - **ACTION: @davidsint**
- Back end communication
  - Web sockets
  - Event sourcing
  - REST
  - GraphQL
  - Cloud Events
  - Waiting on responses
- Testing
- Storage
  - Local Persistence (Index DB, localStorage)
- Internationalization
- Accessibility
- Monorepos
- State management
- Debugging
- Performance - **ACTION: @FrivalszkyP** 
- Cache
- Environment variables
- Front end frameworks
- Error handling
  - Promises async/await
- Observability
  - Logging/debugging, forward trace id to backend
  - Metrics ?
  - Tracing
- Routing
- Graphics, Visualization, Charts
- 3D modeling
- Cross platform (Web, mobile, desktop)
  - PWAs
  - Hybrid apps (cordova, phonegap etc.)
  - React Native
  - WebViews, Chrome Custom Tabs, Trusted Web Activities
  - Thick versus thin clients
  - Electron
- Offline / Service Workers / Web Workers
- Project layout
	- `.npmrc`
	- `.nvmrc`
- static code analysis and tooling - **ACTION: @ctcpip** 
	- eslint
	- prettier
	- editorconfig

**Comment by ctcpip** [comment]
I'll take a first pass at `static code analysis and tooling`

**Comment by tomdev10** [comment]
Hi - I am happy to take "Other Rendering Methods", if I could be assigned please @mhdawson - thanks

**Comment by mhdawson** [comment]
@tomdev10 done.

**Comment by lholmquist** [comment]
Wanted to add an update here.  michael and I went through the list above and created some higher level headings that are in line with the node reference architecture and i've added that to the readme:  https://github.com/nodeshift/web-application-reference#components

We've structured the list in such a way that we now have a "common", "backend" and "front-end" sections under each of the main headings.  And if we find that an section under common is different enough we can move it.  If there are slight differences for a component under the common heading, we can add a couple paragraphs.. and so forth
